File: PythonCode/collapsed_particle_gibbs.py
######################################################################################
#
# Author : Niharika Gauraha
#          KTH
#          Email : niharika@kth.se
#
# Published under GNU General Public License
#####################################################################################
# Implements Collapsed particle Gibbs

import numpy as np
import utils
import sys
import scipy.stats as stats
from marg_csmc import MargCSMC


class CollapsedPG(MargCSMC):
    def __init__(self, f, g, prior):
        """ This method is called when you create an instance of the class."""
        # Stop, if input parameters are NULL
        if (f is None) or (g is None):
            sys.exit("Error: the input parameters are NULL")
        self._q_est = None
        self._r_est = None
        self._x_est = None
        MargCSMC.__init__(self, f, g, prior)

    def generate_samples(self, y, x0, q_init, r_init, prior_a, prior_b, n_particles=100, max_iter=100):
        # Stop, if input parameters are NULL
        if y is None:
            sys.exit("Error: the input parameters are NULL")

        self._seq_len = len(y)
        q = np.zeros(max_iter)
        r = np.zeros(max_iter)
        x = np.zeros((max_iter, self._seq_len))

        q[0] = q_init
        r[0] = r_init
        x_ref = np.zeros(self._seq_len)

        # Initialize the state by running a CPF
        x[0, :] = self.sample_states(y, x0, x_ref, n_particles)

        # Run MCMC loop
        for k in range(1, max_iter):
            # Sample the parameters (inverse gamma posteriors)
            seq_t_1 = np.array(range(0, self._seq_len-1))
            err_q = x[k-1, 1:self._seq_len] - self._f(x[k-1, seq_t_1], seq_t_1)
            err_q = sum(err_q**2)
            q[k] = stats.invgamma.rvs(a=prior_a + (self._seq_len-1)/2, scale=(prior_b + err_q/2), size=1)

            err_r = y - self._g(x[k-1, :])
            err_r = sum(err_r**2)
            r[k] = stats.invgamma.rvs(a=prior_a + self._seq_len/2, scale=(prior_b + err_r/2), size=1)

            x[k, :] = self.sample_states(y, x0, x[k-1, :], n_particles, max_iter=1)

        self._q_est = q
        self._r_est = r
        self._x_est = x

    def get_states(self):
        if self._x_est is None:
            sys.exit("call simulate method first")
        return self._x_est[-1]

    def get_q(self):
        if self._q_est is None:
            sys.exit("call simulate method first")
        return self._q_est

    def get_r(self):
        if self._r_est is None:
            sys.exit("call simulate method first")
        return self._r_est


if __name__ == '__main__':
    from matplotlib import pyplot as plt
    # Set up some parameters
    n_particle = 100  # Number of particles
    seq_len = 100  # Length of data record
    f1 = utils.state_transition_func
    g1 = utils.transfer_func
    # R = 1.
    # Q = 0.1
    xref = np.zeros(seq_len)
    # Generate data
    data_x, data_y = utils.generate_data(r=1.0, q=0.1, seq_len=seq_len)

    rinit = 0.1
    qinit = 1.
    priora = .01
    priorb = .01
    n_iter = 10
    burnIn = int(n_iter*.3)
    pg = CollapsedPG(f1, g1, np.ones(4))

    pg.generate_samples(data_y, 0, qinit, rinit, priora, priorb, n_particles=n_particle, max_iter=n_iter)
    q_est = pg.get_q()
    r_est = pg.get_r()
    x_mult = pg.get_states()

    plt.plot(data_x, marker='o', label='Real States', markersize=3)
    plt.plot(x_mult, label='CPF+Multinomial Filtered States')
    plt.legend()
    plt.show()

    # plt.show()
    plt.savefig("plots/PG")
    # plt.clf()

    n_bins = int(np.floor(np.sqrt(n_iter - burnIn)))
    grid = np.arange(burnIn, n_iter, 1)

    # Plot Q=q^2 (black line shows the posterior mean)
    q_trace = q_est[burnIn:n_iter]
    plt.subplot(3, 1, 1)
    plt.hist(q_trace, n_bins, density=True, facecolor='#9670B3')
    plt.xlabel("Q")
    plt.ylabel("Estimated posterior density")
    plt.axvline(np.mean(q_trace), color='k')

    # Plot the trace of the Markov chain after burn-in (solid black line = posterior mean)
    plt.subplot(3, 1, 2)
    plt.plot(grid, q_trace, color='#9670B3')
    plt.xlabel("iteration")
    plt.ylabel("Q")
    plt.axhline(np.mean(q_trace), color='k')

    # Plot the autocorrelation function
    plt.subplot(3, 1, 3)
    macf = np.correlate(q_trace - np.mean(q_trace), q_trace - np.mean(q_trace), mode='full')
    idx = int(macf.size / 2)
    macf = macf[idx:]
    macf = macf[0:100]
    macf /= macf[0]
    grid = range(len(macf))
    plt.plot(grid, macf, color='#9670B3')
    plt.xlabel("lag")
    plt.ylabel("ACF of Q")

    plt.show()

    # Plot R=r^2 (black line shows the posterior mean)
    r_trace = q_est[burnIn:n_iter]
    plt.subplot(3, 1, 1)
    plt.hist(r_trace, n_bins, density=True, facecolor='#708CB3')
    plt.xlabel("R")
    plt.ylabel("Estimated posterior density")
    plt.axvline(np.mean(r_trace), color='k')

    # Plot the trace of the Markov chain after burn-in (solid black line = posterior mean)
    plt.subplot(3, 1, 2)
    plt.plot(grid, r_trace, color='#708CB3')
    plt.xlabel("iteration")
    plt.ylabel("R")
    plt.axhline(np.mean(r_trace), color='k')

    # Plot the autocorrelation function
    plt.subplot(3, 1, 3)
    macf = np.correlate(r_trace - np.mean(r_trace), r_trace - np.mean(r_trace), mode='full')
    idx = int(macf.size / 2)
    macf = macf[idx:]
    macf = macf[0:100]
    macf /= macf[0]
    grid = range(len(macf))
    plt.plot(grid, macf, color='#708CB3')
    plt.xlabel("lag")
    plt.ylabel("ACF of R")

    plt.show()

File: RustCode/src/collapsed_particle_gibbs.rs
use std::process::exit;
use rand::{SeedableRng};
use rand::rngs::StdRng;
use rand::distributions::Distribution;
use statrs::distribution::{InverseGamma};
use crate::marg_csmc;
use crate::utils;

type FDef = fn(&Vec<f64>, usize) -> Vec<f64>;
type GDef = fn(&Vec<f64>) -> Vec<f64>;

// conditional_SMC class
#[allow(dead_code)]
pub struct CollapsedPG {
    _q_est: Vec<f64>,
    _r_est: Vec<f64>,
    _x_est: Vec<f64>,
    _f: FDef,
    _g: GDef,
    _rng: StdRng,
    _mcsmc:marg_csmc::MargCSMC
}

#[allow(dead_code)]
impl CollapsedPG {
    pub fn new(f: FDef, g: GDef) -> Self {
        CollapsedPG {
            _q_est: Default::default(),
            _r_est: Default::default(),
            _x_est: Default::default(),
            _f: f,
            _g:g,
            _rng: StdRng::from_entropy(),
            _mcsmc: marg_csmc::MargCSMC::new(f, g, marg_csmc::Priors(1.0,1.0,1.0,1.0))
        }
    }

    // returns the last state trajectory
    pub fn get_states(&mut self) -> Vec<f64> {
        if self._x_est.len() == 0 {
            println!("call generateWeightedParticles first");
            exit(0);
        }
        self._x_est.clone()
    }
    // returns estimated q
    pub fn get_q(&mut self) -> Vec<f64> {
        if self._q_est.len() == 0 {
            println!("call generateWeightedParticles first");
            exit(0);
        }
        self._q_est.clone()
    }

    // returns estimated r
    pub fn get_r(&mut self) -> Vec<f64> {
        if self._r_est.len() == 0 {
            println!("call generateWeightedParticles first");
            exit(0);
        }
        self._r_est.clone()
    }

    pub fn generate_samples(&mut self, y: &Vec<f64>, x0: f64, x_ref: &Vec<f64>,
        q_init: f64, r_init: f64, prior_a: f64, prior_b: f64, n_particles:usize, max_iter:usize){
        let seq_len:usize = y.len();
        self._q_est = vec![0.0; max_iter];
        self._r_est = vec![0.0; max_iter];
        self._x_est = vec![0.0; seq_len];
        self._q_est[0] = q_init;
        self._r_est[0] = r_init;

        self._x_est = self._mcsmc.sample_states(y, x0, x_ref, n_particles, 1);

        for iter in 1..max_iter{
            let mut err_q:f64 = 0.0;

            for t in 1..seq_len {
                let xpred = utils::_state_transition_func(self._x_est[t-1], t-1);
                err_q += (xpred - self._x_est[t]).powi(2);
            }
            let n = InverseGamma::new( prior_a + ((seq_len-1) as f64)/2.0,
                                       prior_b + err_q/2.0 ).unwrap();
            self._q_est[iter] = n.sample(&mut self._rng);
            //println!("{}", n.sample(&mut self._rng));
            let mut err_r:f64 = 0.0;

            for t in 0..seq_len {
                let ypred = utils::_transfer_func(self._x_est[t]);
                err_r += (ypred - y[t]).powi(2);
            }

            let n = InverseGamma::new( prior_a + (seq_len as f64)/2.0,
                                           prior_b + err_r/2.0 ).unwrap();
            self._r_est[iter] = n.sample(&mut self._rng);

            self._x_est = self._mcsmc.sample_states(y, x0, &self._x_est, n_particles, 1);
        }
        println!("hello");
    }
}